src/ContentCreator.php

```diff
@@ -0,0 +1,858 @@
+      $plugin = $this->baseEntityTablesPluginManager->createInstance($entity_type, $this->config[$entity_type]);
+      $this->cacheBaseEntityTablesPlugin[$entity_type] = $plugin;
+
+      return $this->cacheBaseEntityTablesPlugin[$entity_type];
+    }
+
+    $plugin = $this->baseEntityTablesPluginManager->createInstance('generic', $this->config[$entity_type]);
+    $this->cacheBaseEntityTablesPlugin[$entity_type] = $plugin;
+
+    return $this->cacheBaseEntityTablesPlugin[$entity_type];
+  }
+
+  /**
+   * Return row templates for base tables.
+   *
+   * @param string $entity_type
+   *   The entity type.
+   *
+   * @return array
+   *   Returns row templates for base tables.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   */
+  protected function getBaseTableTemplates($entity_type) {
+    if (isset($this->cacheBaseTableRowTemplates[$entity_type])) {
+      return $this->cacheBaseTableRowTemplates[$entity_type];
+    }
+
+    $base_entity_tables_plugin = $this->getBaseEntityTablesPlugin($entity_type);
+    $this->cacheBaseTableRowTemplates[$entity_type] = $base_entity_tables_plugin->getBaseTableTemplates();
+
+    return $this->cacheBaseTableRowTemplates[$entity_type];
+  }
+
+  /**
+   * Creates CSV file for database import.
+   *
+   * @param string $entity_type
+   *   The entity type.
+   * @param string $bundle_type
+   *   The bundle type.
+   * @param int $num_of_instances
+   *   The number of instances for bundle.
+   * @param int $parent_id
+   *   The parent entity id.
+   * @param string $parent_type
+   *   The parent entity type.
+   * @param string $parent_field_name
+   *   The parent field name with references to new created entities.
+   *
+   * @return int
+   *   Returns number of created entities.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  protected function createBundle($entity_type, $bundle_type, $num_of_instances, $parent_id = NULL, $parent_type = NULL, $parent_field_name = NULL) {
+    $unique_bundle_key = $entity_type . '_' . $bundle_type;
+    if (isset($this->entityTypeReferenceNestingStack[$unique_bundle_key])) {
+      return 0;
+    }
+
+    $entity_type_state = [
+      'entity_type' => $entity_type,
+      'bundle_type' => $bundle_type,
+      'parent_id' => $parent_id,
+      'parent_type' => $parent_type,
+      'parent_field_name' => $parent_field_name,
+    ];
+
+    $this->entityTypeReferenceNestingStack[$unique_bundle_key] = $entity_type_state;
+
+    $this->initBaseEntityTables($entity_type);
+
+    $entity_type_index = $this->globalState['entity_states'][$entity_type]['index'];
+    $bundle_type_index = $this->globalState['entity_states'][$entity_type][$bundle_type]['index'];
+    $entity_type_total = $this->globalState['entity_states'][$entity_type]['count'];
+
```

Comment:
Perhaps we could get rid of all the `index` keys by doing something like `$entity_type_uuid_part = array_search($entity_type, array_keys($this->globalState['entity_states']));`

Reply:
To be honest, I would avoid as much as possible `array_search` and other `array_*` functions. Main reason for that is performance and we need that in this case. I think that "direct access" is better in this case.

Bad thing is also too much nesting, maybe I could split `globalState` into multiple variables.

Reply:
Are you sure that this is not premature optimisation?

Reply:
@alexpott I have tried and it looks like Mr. Knuth knows what he is talking about!